compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnvilModuleDescriptor.kt

```diff
@@ -23,47 +22,50 @@ public interface AnvilModuleDescriptor : ModuleDescriptor {
     lookupLocation: LookupLocation = FROM_BACKEND
   ): ClassDescriptor?
 
-  public fun getClassesAndInnerClasses(ktFile: KtFile): List<KtClassOrObject>
-
-  public fun getKtClassOrObjectOrNull(fqName: FqName): KtClassOrObject?
+  public fun getClassAndInnerClassReferences(ktFile: KtFile): List<Psi>
 
   public fun getClassReference(clazz: KtClassOrObject): Psi
 
   public fun getClassReference(descriptor: ClassDescriptor): Descriptor
 
-  public fun getClassReferenceOrNull(fqName: FqName): ClassReference?
+  /**
+   * Attempts to resolve the [FqName] to a [ClassDescriptor] first, then falls back to a
+   * [KtClassOrObject] if the descriptor resolution fails. This will happen if the code being
+   * parsed was generated as part of the compilation round for this module.
+   *
+   * If [preferDescriptor] is `false`, then the order of resolution is reverted.
+   */
+  public fun getClassReferenceOrNull(
+    fqName: FqName,
+    preferDescriptor: Boolean = true
+  ): ClassReference?
 }
 
 @Suppress("NOTHING_TO_INLINE")
 internal inline fun ModuleDescriptor.asAnvilModuleDescriptor(): AnvilModuleDescriptor =
   this as AnvilModuleDescriptor
 
-@ExperimentalAnvilApi
-public fun FqName.getKtClassOrObjectOrNull(
-  module: ModuleDescriptor
-): KtClassOrObject? = module.asAnvilModuleDescriptor().getKtClassOrObjectOrNull(this)
-
-@ExperimentalAnvilApi
-public fun FqName.getKtClassOrObject(
-  module: ModuleDescriptor
-): KtClassOrObject = getKtClassOrObjectOrNull(module)
-  ?: throw AnvilCompilationException("Couldn't resolve KtClassOrObject for ${asString()}")
-
 @ExperimentalAnvilApi
 public fun FqName.canResolveFqName(
   module: ModuleDescriptor
 ): Boolean = module.asAnvilModuleDescriptor().resolveClassIdOrNull(classIdBestGuess()) != null
 
 @ExperimentalAnvilApi
-public fun KtFile.classesAndInnerClasses(
+@Deprecated(
+  message = "Use ClassReference instead.",
+  replaceWith = ReplaceWith("classAndInnerClassReferences(module)")
```

Comment:
This is now deprecated and we need to migrate all `CodeGenerator`s to this new API. That's where help is appreciated :-)